scipy/sparse/linalg/eigen/tests/test_svds.py

```diff
@@ -223,24 +83,404 @@ def _rmatvec(self, x):
+            error = err(tol)
+            assert error < accuracy
+            assert error > accuracy/10
+
+    def test_svd_v0(self):
+        # check that the `v0` parameter affects the solution
+        n = 100
+        k = 1
+        # If k != 1, LOBPCG needs more initial vectors, which are generated
+        # with random_state, so it does not pass w/ k >= 2.
+        # For some other values of `n`, the AssertionErrors are not raised
+        # with different v0s, which is reasonable.
+
+        rng = np.random.default_rng(0)
+        A = rng.random((n, n))
+
+        # with the same v0, solutions are the same, and they are accurate
+        # v0 takes precedence over random_state
+        v0a = rng.random(n)
+        res1a = svds(A, k, v0=v0a, solver=self.solver, random_state=0)
+        res2a = svds(A, k, v0=v0a, solver=self.solver, random_state=1)
+        assert_equal(res1a, res2a)
+        _check_svds(A, k, *res1a)
+
+        # with the same v0, solutions are the same, and they are accurate
+        v0b = rng.random(n)
+        res1b = svds(A, k, v0=v0b, solver=self.solver, random_state=2)
+        res2b = svds(A, k, v0=v0b, solver=self.solver, random_state=3)
+        assert_equal(res1b, res2b)
+        _check_svds(A, k, *res1b)
+
+        # with different v0, solutions can be numerically different
+        message = "Arrays are not equal"
+        with pytest.raises(AssertionError, match=message):
+            assert_equal(res1a, res1b)
+
+    def test_svd_random_state(self):
+        # check that the `random_state` parameter affects the solution
+        # Admittedly, `n` and `k` are chosen so that all solver pass all
+        # these checks. That's a tall order, since LOBPCG doesn't want to
+        # achieve the desired accuracy and ARPACK often returns the same
+        # singular values/vectors for different v0.
+        n = 100
+        k = 1
+
+        rng = np.random.default_rng(0)
+        A = rng.random((n, n))
+
+        # with the same random_state, solutions are the same and accurate
+        res1a = svds(A, k, solver=self.solver, random_state=0)
+        res2a = svds(A, k, solver=self.solver, random_state=0)
+        assert_equal(res1a, res2a)
+        _check_svds(A, k, *res1a)
+
+        # with the same random_state, solutions are the same and accurate
+        res1b = svds(A, k, solver=self.solver, random_state=1)
+        res2b = svds(A, k, solver=self.solver, random_state=1)
+        assert_equal(res1b, res2b)
+        _check_svds(A, k, *res1b)
+
+        # with different random_state, solutions can be numerically different
+        message = "Arrays are not equal"
+        with pytest.raises(AssertionError, match=message):
+            assert_equal(res1a, res1b)
+
+    @pytest.mark.parametrize("random_state", (0, 1,
+                                              np.random.RandomState(0),
+                                              np.random.default_rng(0)))
+    def test_svd_random_state_2(self, random_state):
```

Comment:
Do we need this test? Looks like we are testing `check_random_state` instead.

Reply:
Knowledge of the innards of a function may inspire new tests to write, but I'm not sure that it's always good practice to limit the tests based on how we think the function is working. That is, I know that `check_random_state` is used to standardize the `random_state` instance, so everything _should_ work correctly for any valid `random_state`, but I have made mistakes before. If this test does not take long to execute, I think it is worth the sanity check - especially given the nonstandard use of random numbers in `method='arpack'`/`method='lobpcg'` and the fact the Fortran routines underlying these methods seem to use their own RNGs under certain circumstances (which we want to avoid).

The primary purpose of `test_svd_random_state` is to check that the use of a different `random_state` object _does_ affect the solution.
The primary purpose of `test_svd_random_state_3` is to check that a given `random_state` object in a different internal state also affects the solution.
The primary purpose of `test_svd_random_state_2` is to check that `random_state` works correctly for each `random_state` type.

There is some overlap, but they are intended to be distinct tests. Perhaps not all functions are (or need to be) tested like this, but I tried to be thorough in `test_svds` and didn't stop myself from writing tests I thought might fail.

If this makes sense, I can add these notes as comments in the tests.

Reply:
If the RNG's type could make a difference for the underlying methods then ok. As you said it's fast so not a big deal. It's fine without the comment in the code, the tests are simple enough to understand quickly. I just wanted to make sure we were not testing `check_random_state`